osdep/macosx_touchbar.m

```diff
@@ -0,0 +1,202 @@
+                               name:@"Previous Chapter"
+                              image:NSImageNameTouchBarSkipBackTemplate
+                             action:@selector(previousChapter:)];
+    } else if ([identifier isEqualToString:nextChapter]) {
+        return [self tbButtonWithId:identifier
+                               name:@"Next Chapter"
+                              image:NSImageNameTouchBarSkipAheadTemplate
+                             action:@selector(nextChapter:)];
+    } else if ([identifier isEqualToString:cycleAudio]) {
+        return [self tbButtonWithId:identifier
+                               name:@"Cycle Audio"
+                              image:NSImageNameTouchBarAudioInputTemplate
+                             action:@selector(cycleAudio:)];
+    } else if ([identifier isEqualToString:cycleSubtitle]) {
+        return [self tbButtonWithId:identifier
+                               name:@"Cycle Subtitle"
+                              image:NSImageNameTouchBarComposeTemplate
+                             action:@selector(cycleSubtitle:)];
+    } else if ([identifier isEqualToString:currentPosition]) {
+        NSCustomTouchBarItem *tbItem = [self tbTextWithId:identifier name:@"Current Position"];
+        self.touchbarCurrentPosition = (NSTextField *)tbItem.view;
+        return tbItem;
+    } else if ([identifier isEqualToString:timeLeft]) {
+        NSCustomTouchBarItem *tbItem = [self tbTextWithId:identifier name:@"Time Left"];
+        self.touchbarTimeLeft = (NSTextField *)tbItem.view;
+        return tbItem;
+    }
+    return nil;
+}
+
+- (NSCustomTouchBarItem *)tbButtonWithId:(NSTouchBarItemIdentifier) identifier
+    name:(NSString *)label image:(NSString *)image action:(SEL)action
+{
+    NSCustomTouchBarItem *tbItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:identifier];
+    NSImage *tbImage = [NSImage imageNamed:image];
+    tbItem.view = [NSButton buttonWithImage:tbImage target:self action:action];
+    tbItem.customizationLabel = label;
+    return tbItem;
+}
+
+- (NSCustomTouchBarItem *)tbTextWithId:(NSTouchBarItemIdentifier) identifier
+                                  name:(NSString *)label
+{
+    NSCustomTouchBarItem *tbItem = [[NSCustomTouchBarItem alloc] initWithIdentifier:identifier];
+    NSTextField *tbText = [NSTextField labelWithString:@"0:00"];
+    tbText.alignment = NSTextAlignmentCenter;
+    tbItem.view = tbText;
+    tbItem.customizationLabel = label;
+    return tbItem;
+}
+
+- (NSString *)formatTimeFrom:(int)time
+{
+    int seconds = time % 60;
+    int minutes = (time / 60) % 60;
+    int hours = time / (60 * 60);
+
+    NSString *stime = hours > 0 ? [NSString stringWithFormat:@"%d:", hours] : @"";
+    stime = (stime.length > 0 || minutes > 9) ?
+        [NSString stringWithFormat:@"%@%02d:", stime, minutes] :
+        [NSString stringWithFormat:@"%d:", minutes];
+    stime = [NSString stringWithFormat:@"%@%02d", stime, seconds];
```

Comment:
yes, i first wanted to use "playback-time" but it always returns HH:MM:SS. the space on the touchbar can be a bit sparse, depending on how many items you want on it, so i decided it would be a good idea to fall back to only (M)M:SS when (H)H is not needed. also omitting a useless leading 0.